### ISSUE TITLE
hnswlib dependency issue

### ISSUE THREAD

**Issue by HarshalRathore** [nicknochnack/LangchainDocuments#1]
**Which package caused the issue?**
[hnswlib ](https://github.com/nmslib/hnswlib)

**What was the issue** 
I think it requires Clang with version `14>` which can be installed with installing Microsoft Visual Studio.

**Software/OS/platform with versions used** 
```
Windows 10 Home
Python 3.10.9
pip 23.0.1
```

**Complete Error** 
```
Building wheels for collected packages: hnswlib, sentence-transformers, validators
Building wheel for hnswlib (pyproject.toml) ... error
error: subprocess-exited-with-error

× Building wheel for hnswlib (pyproject.toml) did not run successfully.
│ exit code: 1
╰─> [5 lines of output]
   running bdist_wheel
   running build
   running build_ext
   building 'hnswlib' extension
   error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
   [end of output]

note: This error originates from a subprocess and is likely not a problem with pip.
ERROR: Failed building wheel for hnswlib
Building wheel for sentence-transformers (setup.py) ... done
Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125961 sha256=e57909d0f8487ab4d6ed17e8230321d2419f34de1f3347d6ddfa1ed8ab8b54a8
Stored in directory: c:\users\harshal rathore\appdata\local\pip\cache\wheels\62\f2\10\1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Building wheel for validators (setup.py) ... done
Created wheel for validators: filename=validators-0.20.0-py3-none-any.whl size=19590 sha256=0e8482e82775ea9a61821e0b79d086368ce3ab7d0420a01e01440c641ffc1ead
Stored in directory: c:\users\harshal rathore\appdata\local\pip\cache\wheels\f2\ed\dd\d3a556ad245ef9dc570c6bcd2f22886d17b0b408dd3bbb9ac3
Successfully built sentence-transformers validators
Failed to build hnswlib
ERROR: Could not build wheels for hnswlib, which is required to install pyproject.toml-based projects

```

**Comment by shihaozhangjohn** [comment]
Running on Mac and tried everything i could ,.. failed

**Comment by HarshalRathore** [comment]
> Running on Mac and tried everything i could ,.. failed

One of the package in requirement.txt is pywin32 which is a window specific pkg so you cannot use this code on any Unix platform

**Comment by shihaozhangjohn** [comment]
> > Running on Mac and tried everything i could ,.. failed
> 
> One of the package in requirement.txt is pywin32 which is a window specific pkg so you cannot use this code on any Unix platform
> 
> @nicknochnack can you please resolve these issues.

yeah i deleted that one from requirement txt, and i tried python 3.10.10 and everything i could find on google/stackoverflow and etc. just cannot install chromadb...sad.... i guess it's due to clang or c++ complier . not sure

**Comment by wellbranch** [comment]
> > > Running on Mac and tried everything i could ,.. failed
> > 
> > 
> > One of the package in requirement.txt is pywin32 which is a window specific pkg so you cannot use this code on any Unix platform
> > @nicknochnack can you please resolve these issues.
> 
> yeah i deleted that one from requirement txt, and i tried python 3.10.10 and everything i could find on google/stackoverflow and etc. just cannot install chromadb...sad.... i guess it's due to clang or c++ complier . not sure

I am on Mac and got stuck on this too. After some testing I found a way that worked with some help from Stackoverflow.

**Think these were the only steps I needed to get it working:**
1. Removed "pywin32" from requirements.txt
2. Wrote "export HNSWLIB_NO_NATIVE=1" in the terminal (to be able to install chromadb)
3. Then re-ran the app with "streamlit run app.py"

**Comment by shihaozhangjohn** [comment]
> 

thanks for ur kindly reply,, i tried step2 and could not got chromadb installed lol

**Comment by shihaozhangjohn** [comment]
> > > > Running on Mac and tried everything i could ,.. failed
> > > 
> > > 
> > > One of the package in requirement.txt is pywin32 which is a window specific pkg so you cannot use this code on any Unix platform
> > > @nicknochnack can you please resolve these issues.
> > 
> > 
> > yeah i deleted that one from requirement txt, and i tried python 3.10.10 and everything i could find on google/stackoverflow and etc. just cannot install chromadb...sad.... i guess it's due to clang or c++ complier . not sure
> 
> I am on Mac and got stuck on this too. After some testing I found a way that worked with some help from Stackoverflow.
> 
> **Think these were the only steps I needed to get it working:**
> 
> 1. Removed "pywin32" from requirements.txt
> 2. Wrote "export HNSWLIB_NO_NATIVE=1" in the terminal (to be able to install chromadb)
> 3. Re-run the requirements install with "pip install -r requirements.txt"
> 4. Run the app with "streamlit run app.py"


thanks for ur kindly reply,, i tried step2 and could not got chromadb installed lol

**Comment by wellbranch** [comment]
> > > > > Running on Mac and tried everything i could ,.. failed
> > > > 
> > > > 
> > > > One of the package in requirement.txt is pywin32 which is a window specific pkg so you cannot use this code on any Unix platform
> > > > @nicknochnack can you please resolve these issues.
> > > 
> > > 
> > > yeah i deleted that one from requirement txt, and i tried python 3.10.10 and everything i could find on google/stackoverflow and etc. just cannot install chromadb...sad.... i guess it's due to clang or c++ complier . not sure
> > 
> > 
> > I am on Mac and got stuck on this too. After some testing I found a way that worked with some help from Stackoverflow.
> > **Think these were the only steps I needed to get it working:**
> > 
> > 1. Removed "pywin32" from requirements.txt
> > 2. Wrote "export HNSWLIB_NO_NATIVE=1" in the terminal (to be able to install chromadb)
> > 3. Re-run the requirements install with "pip install -r requirements.txt"
> > 4. Run the app with "streamlit run app.py"
> 
> thanks for ur kindly reply,, i tried step2 and could not got chromadb installed lol

Okay, sorry to hear that. Is it the chromadb install keeps failing if you do "pip install chromadb"? Or does it get stuck somewhere else in the requirements install? I realise now that I also removed "PyYAML" from requirements.txt per the suggestion in [this pull request (https://github.com/nicknochnack/LangchainDocuments/pull/3/commits/8a153f4a0165c97eb12fb301e82ab80848b94f3e), not sure if it has an impact

**Comment by shihaozhangjohn** [comment]
> > > > > > Running on Mac and tried everything i could ,.. failed
> > > > > 
> > > > > 
> > > > > One of the package in requirement.txt is pywin32 which is a window specific pkg so you cannot use this code on any Unix platform
> > > > > @nicknochnack can you please resolve these issues.
> > > > 
> > > > 
> > > > yeah i deleted that one from requirement txt, and i tried python 3.10.10 and everything i could find on google/stackoverflow and etc. just cannot install chromadb...sad.... i guess it's due to clang or c++ complier . not sure
> > > 
> > > 
> > > I am on Mac and got stuck on this too. After some testing I found a way that worked with some help from Stackoverflow.
> > > **Think these were the only steps I needed to get it working:**
> > > 
> > > 1. Removed "pywin32" from requirements.txt
> > > 2. Wrote "export HNSWLIB_NO_NATIVE=1" in the terminal (to be able to install chromadb)
> > > 3. Re-run the requirements install with "pip install -r requirements.txt"
> > > 4. Run the app with "streamlit run app.py"
> > 
> > 
> > thanks for ur kindly reply,, i tried step2 and could not got chromadb installed lol
> 
> Okay, sorry to hear that. Is it the chromadb install keeps failing if you do "pip install chromadb"? Or does it get stuck somewhere else in the requirements install? I realise now that I also removed "PyYAML" from requirements.txt per the suggestion in [this pull request](https://github.com/nicknochnack/LangchainDocuments/pull/3/commits/8a153f4a0165c97eb12fb301e82ab80848b94f3e), not sure if it has an impact

no luck lol, i guess that was still the clang issue. my python version is 3.9.16 on Monterey 12.6.6, with xcode installed, gcc installed and etc.  i found that err msg like " error: command '/usr/bin/clang' failed with exit code 1" and 'ERROR: Could not build wheels for hnswlib, which is required to install pyproject.toml-based projects' appeared for these kind of issue...

I'm thinking to update my MacOs for next try.

**Comment by linviaz** [comment]
If your python is from a conda environment, `conda install -c conda-forge hnswlib` works to install hnswlib.
